lib/timers.js

```diff
@@ -10,36 +10,84 @@ const kOnTimeout = Timer.kOnTimeout | 0;
-// we'll use a single watcher for all sockets with the same timeout value
-// and a linked list. This technique is described in the libev manual:
+// To achieve a high level of efficiency, the implementation is as distributed
+// and lazy as possible.
+//
+// Object maps are kept which contain linked lists keyed by their duration in
+// milliseconds.
+// The linked lists within are TimerWrap C++ handles with a linked list appended
+// to their JavaScript representation.
+//
+// With this, each list of timers with the same duration reduces the overhead of
+// trying to track when all of the timers should timeout by only tracking the
+// first timeout, and adjusting to sooner or later timeouts when applicable.
+// This technique is descirbed in the libev manual:
 // http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#Be_smart_about_timeouts
+//
+// The timers are distributed into separate lists by milliseconds primarly to
+// reduce the impact of the linked list's O(n) (linear) insertion time.
 
-// Object containing all lists, timers
-// key = time in milliseconds
-// value = list
-var lists = {};
 
+// Object maps containing linked lists of timers, keyed and sorted by their
+// duration in miliseconds.
+// - key = time in milliseconds
+// - value = linked list
+const refedLists = {};
+const unrefedLists = {};
+// The difference between these two objects is that the former contains timers
+// that will keep the process open if they are the only thing left, while the
+// latter will not.
 
-// call this whenever the item is active (not idle)
-// it will reset its timeout.
-// the main function - creates lists on demand and the watchers associated
-// with them.
+
+// Schedule or re-schedule a timer.
+// Call this whenever the item is active (not idle).
+// The item must have been enroll()'d first.
 exports.active = function(item) {
+  insert(item, false);
+};
+
+// Internal APIs that need timeouts should use `_unrefActive()` instead of
+// `active()` so that they do not unnecessarily keep the process open.
+exports._unrefActive = function(item) {
+  insert(item, true);
+};
+
+
+// The underlying logic for scheduling or re-scheduling a timer.
+//
+// Inserts a timer into an existing timers list, or creates a new TimerWrap
+// backed list if one does not already exist for the specified timeout duration.
+function insert(item, unrefed) {
   const msecs = item._idleTimeout;
   if (msecs < 0 || msecs === undefined) return;
 
   item._idleStart = Timer.now();
 
   var list;
+  const lists = unrefed ? unrefedLists : refedLists;
 
+  // Use an existing list of there is one, otherwise we need to make a new one.
   if (lists[msecs]) {
     list = lists[msecs];
   } else {
+    // Make a new linked list of timers.
+    // The list object is a TimerWrap, a C++ handle backing the timers for
+    // efficiency, and the linked list is appended onto it.
     list = new Timer();
+    if (unrefed) list.unref();
+    list._unrefed = unrefed;
```

Comment:
@Fishrock123 yes `_parent` sounds good, then down below you can do `delete list._parent[msecs];` instead of
```
if (list._unrefed) {  
  delete unrefedLists[msecs];  
} else {  
  delete refedLists[msecs];  
}  
```